substrate/frame/support/src/utils.rs

```diff
@@ -0,0 +1,99 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::Parameter;
+use codec::{Decode, Encode, Input, MaxEncodedLen};
+use frame_support_procedural::{CloneNoBound, EqNoBound, PartialEqNoBound, RuntimeDebugNoBound};
+use scale_info::TypeInfo;
+use sp_core::Get;
+
+/// TODO: docs
+pub struct InputWithMaxLength<'a, I: Input, S: Get<usize>> {
+	input: &'a mut I,
+	remaining_len: usize,
+	_phantom: core::marker::PhantomData<S>,
+}
+
+impl<'a, I: Input, S: Get<usize>> InputWithMaxLength<'a, I, S> {
+	pub fn new(input: &'a mut I) -> Self {
+		Self { input, remaining_len: S::get(), _phantom: Default::default() }
+	}
+}
+
+impl<'a, I: Input, S: Get<usize>> Input for InputWithMaxLength<'a, I, S> {
+	fn remaining_len(&mut self) -> Result<Option<usize>, codec::Error> {
+		self.input.remaining_len().map(|l| l.map(|l| l.min(self.remaining_len)))
+	}
+
+	fn read(&mut self, into: &mut [u8]) -> Result<(), codec::Error> {
+		if into.len() > self.remaining_len {
+			return Err("Not enough data to fill buffer".into());
+		}
+		self.input.read(into)
```

Comment:
```suggestion
		let res = self.input.read(into)?;
		self.remaining_len -= into.len();
		Ok(res)
```

We actually need to remove it from the remaining length :D

Reply:
I really need some tests XD
and do we need to remove it before the read or it wouldn’t matter?

Reply:
I thought about it. If this returns an error, I would assume that we bubble that up any way and the decoding failed.